src/transport/raw/TCP.cpp

```diff
@@ -353,171 +439,243 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+    VerifyOrReturnError(address.GetTransportType() == Transport::Type::kTcp, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(mUsedEndPointCount < mActiveConnectionsSize, CHIP_ERROR_NO_MEMORY);
+
+    char addrStr[Transport::PeerAddress::kMaxToStringSize];
+    address.ToString(addrStr);
+    ChipLogProgress(Inet, "Connecting to peer %s.", addrStr);
+
+    ReturnErrorOnFailure(StartConnect(address, appState, outPeerConnState));
+
+    return CHIP_NO_ERROR;
+}
+
+void TCPBase::TCPDisconnect(const PeerAddress & address)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
     // Closes an existing connection
     for (size_t i = 0; i < mActiveConnectionsSize; i++)
     {
-        if (mActiveConnections[i].InUse())
+        if (mActiveConnections[i].IsConnected())
         {
             Inet::IPAddress ipAddress;
             uint16_t port;
             Inet::InterfaceId interfaceId;
 
-            mActiveConnections[i].mEndPoint->GetPeerInfo(&ipAddress, &port);
-            mActiveConnections[i].mEndPoint->GetInterfaceId(&interfaceId);
-            if (address == PeerAddress::TCP(ipAddress, port, interfaceId))
+            err = mActiveConnections[i].mEndPoint->GetPeerInfo(&ipAddress, &port);
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(Inet, "GetPeerInfo error: %" CHIP_ERROR_FORMAT, err.Format());
+                return;
+            }
+
+            err = mActiveConnections[i].mEndPoint->GetInterfaceId(&interfaceId);
+            if (err != CHIP_NO_ERROR)
             {
+                ChipLogError(Inet, "GetInterfaceId error: %" CHIP_ERROR_FORMAT, err.Format());
+                return;
+            }
+            // if (address == PeerAddress::TCP(ipAddress, port, interfaceId))
+            if (ipAddress == address.GetIPAddress() && port == address.GetPort())
+            {
+                char addrStr[Transport::PeerAddress::kMaxToStringSize];
+                address.ToString(addrStr);
+                ChipLogProgress(Inet, "Disconnecting with peer %s.", addrStr);
+
                 // NOTE: this leaves the socket in TIME_WAIT.
                 // Calling Abort() would clean it since SO_LINGER would be set to 0,
                 // however this seems not to be useful.
-                mActiveConnections[i].Free();
-                mUsedEndPointCount--;
+                CloseConnectionInternal(&mActiveConnections[i], CHIP_NO_ERROR, SuppressCallback::Yes);
             }
         }
     }
 }
 
-void TCPBase::OnPeerClosed(Inet::TCPEndPoint * endPoint)
+void TCPBase::TCPDisconnect(Transport::ActiveTCPConnectionState * conn, bool shouldAbort)
 {
-    TCPBase * tcp = reinterpret_cast<TCPBase *>(endPoint->mAppState);
 
-    ChipLogProgress(Inet, "Freeing connection: connection closed by peer");
+    if (conn == nullptr)
+    {
+        ChipLogError(Inet, "Failed to Disconnect. Passed in Connection is null.");
+        return;
+    }
 
-    tcp->ReleaseActiveConnection(endPoint);
+    if (conn->IsConnected())
```

Comment:
I am not quite following this part.  Doesn't CASESession call TCPDisconnect on a connection that might still be getting set up?

Reply:
Yes, it does. Refactored to include a check for IsConnecting() as well.